### PR TITLE
[Wasm64] Implement emmalloc

### PR DESCRIPTION
**Pull request by sbc100**
The primary change here is to use `size_t` in a bunch of places that
previously used `uint32_t`.

This change also increases the alignment of `sbrk()` from 4 to
`alignof(maxalign_t)` (8).

As part of this change I also converted all the tracing/logging in
emmalloc form EM_ASM to `_emscripten_out`/`_emscripten_err`.  The latter
two are disigned for low level tracing an should never allocate heap
memory themselves.  The motivation for this is that `EM_ASM` arguments
that are 64-bit are not yet well supported.  At best the arguments would
come through as BigInts and the EM_ASM code would become more complex,
having to handle both cases.

### CODE REVIEW

**Comment by sbc100** [comment]
Depends on #17685

**Comment by sbc100** [comment]
I think perhaps the EM_ASM_MAIN_THREAD stuff was used here because back in the day there were issues with console.log/console.error message not being show when working with workers under node.  I believe that was addressed back in #14805 so it safe to always depend on `out()`/`err()`, and they should be used consistently for low-level logging.

**Review by dschuff — APPROVED** [review]

**Comment by sbc100** [comment]
I ended up having to update a couple more tests..

File: system/lib/emmalloc.c [hunk 1]
```diff
@@ -9,11 +9,12 @@
  *
  * Assumptions:
  *
- *  - Pointers are 32-bit.
  *  - sbrk() is used to claim new memory (sbrk handles geometric/linear
  *  - overallocation growth)
  *  - sbrk() can be used by other code outside emmalloc.
  *  - sbrk() is very fast in most cases (internal wasm call).
+ *  - sbrk() is very fast in most cases (internal wasm call).
```
Comment (juj) [inline system/lib/emmalloc.c:16]: This comment is duplicated for some reason?

File: system/lib/emmalloc.c [hunk 2]
```diff
@@ -225,7 +225,7 @@ an allocation size to the bucket index that should be looked at. The buckets are
   Bucket 62: [100663296, 134217727], range size=33554432
   Bucket 63: 134217728 bytes and larger. */
 static_assert(NUM_FREE_BUCKETS == 64, "Following function is tailored specifically for NUM_FREE_BUCKETS == 64 case");
-static int compute_free_list_bucket(uint32_t allocSize)
+static int compute_free_list_bucket(size_t allocSize)
```
Comment (juj) [inline system/lib/emmalloc.c:228]: The `bucketIndex` computation is likely wrong in 64-bit mode, since `__builtin_clz` will return a value that is much larger than from before, leading to the asserts on lines 233-234 triggering.

File: system/lib/emmalloc.c [hunk 3]
```diff
@@ -264,49 +264,49 @@ static bool region_is_in_use(Region *r)
...
   assert(ptr);
-  assert(HAS_ALIGNMENT(ptr, sizeof(uint32_t)));
-  assert(HAS_ALIGNMENT(size, sizeof(uint32_t)));
+  assert(HAS_ALIGNMENT(ptr, sizeof(size_t)));
+  assert(HAS_ALIGNMENT(size, sizeof(size_t)));
   assert(size >= sizeof(Region));
-  *(uint32_t*)ptr = size;
-  ((uint32_t*)ptr)[(size>>2)-1] = size;
+  *(size_t*)ptr = size;
+  ((size_t*)ptr)[(size/sizeof(size_t))-1] = size;
```
Comment (juj) [inline system/lib/emmalloc.c:298]: Can you instead make this a macro with a 32-bit vs 64-bit `>>2` vs `>>3` behavior? Sure, compilers will certainly optimize a division, but they won't do it in -O0, and e.g. reading disassemblies when running a debug build of emmalloc to detect issues will then use different code.

File: system/lib/emmalloc.c [hunk 4]
```diff
@@ -352,38 +352,36 @@ static void dump_memory_regions()
...
-      MAIN_THREAD_ASYNC_EM_ASM(console.log('Region 0x'+($0>>>0).toString(16)+', size: '+($1>>>0)+' ('+($2?"used":"--FREE--")+')'),
-        r, r->size, region_ceiling_size(r) == r->size);
+      _emscripten_outf("Region %p, size: %zu (%s)", r, r->size, region_ceiling_size(r) == r->size ? "used" : "--FREE--");
 
       assert(debug_region_is_consistent(r));
-      uint32_t sizeFromCeiling = size_of_region_from_ceiling(r);
+      size_t sizeFromCeiling = size_of_region_from_ceiling(r);
       if (sizeFromCeiling != r->size)
-        MAIN_THREAD_ASYNC_EM_ASM(console.log('Corrupt region! Size marker at the end of the region does not match: '+($0>>>0)), sizeFromCeiling);
+        _emscripten_outf("Corrupt region! Size marker at the end of the region does not match: %zu", sizeFromCeiling);
```
Comment (juj) [inline system/lib/emmalloc.c:369]: Noo :( This will pull in printf() facilities into the build, which will explode code size on small programs. Especially for debug builds, which are exactly the ones that are most often being wasm-dissed, so there will be a increase to code size on minimal repros.

The reason for using MAIN_THREAD_ASYNC_EM_ASM() was that web browsers do not provide any guarantees that calling console.log() or console.error() are lock-free, i.e. won't internally hold a lock from the main browser thread, and won't be proxied by the browser to the main thread.

Since these print functions are called potentially from a worker thread, and from inside an emmalloc acquire lock, calling console.log() there could make the main browser thread and the worker thread deadlock.

I would prefer the regular `EM_ASM()` or `EM_JS()` for logging, since it has an absolutely tiny code size foot print. This also produces a static guarantee that these logging functions won't allocate - I don't know if printf() in libc is required to have that guarantee.

File: system/lib/emmalloc.c [hunk 5]
```diff
@@ -865,25 +863,23 @@ void emmalloc_free(void *ptr)
...
-  uint32_t size = region->size;
+  size_t size = region->size;
 #ifdef EMMALLOC_VERBOSE
   if (size < sizeof(Region) || !region_is_in_use(region))
   {
     if (debug_region_is_consistent(region))
-      // LLVM wasm backend bug: cannot use MAIN_THREAD_ASYNC_EM_ASM() here, that generates internal compiler error
-      // Reproducible by running e.g. other.test_alloc_3GB
-      EM_ASM(console.error('Double free at region ptr 0x' + ($0>>>0).toString(16) + ', region->size: 0x' + ($1>>>0).toString(16) + ', region->sizeAtCeiling: 0x' + ($2>>>0).toString(16) + ')'), region, size, region_ceiling_size(region));
+      _emscripten_errf("Double free at region ptr 0x%p, region->size: %zx, region->sizeAtCeiling: %zx", region, size, region_ceiling_size(region));
```
Comment (juj) [inline system/lib/emmalloc.c:880]: Oh I see, in wasm64 mode this LLVM bug would be hit on almost all prints? Does EM_JS() have that issue?

File: system/lib/emmalloc.c (hunk 4 above)
Comment (sbc100) [reply system/lib/emmalloc.c:369]: The idea being having the _emscripten_out and _emscripten_err C functions is so that low level libraries can log stuff without having to roll their own EM_ASM/EM_JS mechanism.

_emscripten_out and _emscripten_err don't necessarily map to console.error and console.log.    For example under node we map them to `stdout.write` and `stderr.write` to avoid the posiblity of loosing message that you mention.  Perhaps on the browser side we should proxy them to main the thread synchronously.

In other words if `_emscripten_out` and `_emscripten_err` need fixing to be relaible then we should fix them, rather than avoiding them in this one place.

Comment (sbc100) [reply system/lib/emmalloc.c:369]: Regarding the code side issue that having native printf adds, I suppose we could instead do the formatting on the JS side?  I'll take a look at that possibility.

File: system/lib/emmalloc.c (hunk 5 above)
Comment (sbc100) [reply system/lib/emmalloc.c:880]: I'll try to reproduce the bug and see if its still there..   I didn't run into that specific bug, but the more general issues with make EM_ASM blocks that work with both 64-bit and 32-bit wasm.

I'll try reverting all the formatting EM_ASM changes in this PR and try to find anther way to make the macros work in both modes.
